streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskSchedulerImpl.scala

```diff
@@ -0,0 +1,178 @@
+    dag.tasks.foreach { params =>
+      val (taskGroupId, taskDescription) = params
+      0.until(taskDescription.parallelism).map((taskIndex: Int) => {
+        val taskId = TaskId(taskGroupId, taskIndex)
+        val locality = taskLocator.locateTask(taskDescription)
+        val taskLaunchData = TaskLaunchData(taskId, taskDescription, dag.subGraph(taskGroupId))
+        tasks += (taskId -> taskLaunchData)
+        scheduleTaskLater(taskLaunchData, locality)
+      })
+    }
+    val nonLocalityQueue = tasksToBeScheduled.getOrElse(NonLocality, mutable.Queue.empty[TaskLaunchData])
+    //reorder the tasks to distributing fairly on workers
+    val taskQueue = nonLocalityQueue.sortBy(_.taskId.index)
+    tasksToBeScheduled += (NonLocality -> taskQueue)
+  }
+
+  def getResourceRequests(): Array[ResourceRequest] ={
+    fetchResourceRequests(fromOneWorker = false)
+  }
+
+  import Relaxation._
+  private def fetchResourceRequests(fromOneWorker: Boolean = false): Array[ResourceRequest] ={
+    var resourceRequests = Array.empty[ResourceRequest]
+    tasksToBeScheduled.foreach(taskWithLocality => {
+      val (locality, tasks) = taskWithLocality
+      LOG.debug(s"locality : $locality, task queue size : ${tasks.size}")
+
+      if (tasks.size > 0) {
+        locality match {
+          case locality: WorkerLocality =>
+            resourceRequests = resourceRequests :+
+              ResourceRequest(Resource(tasks.size), locality.workerId, relaxation = SPECIFICWORKER)
+          case _ if fromOneWorker =>
+              resourceRequests = resourceRequests :+ ResourceRequest(Resource(tasks.size), relaxation = ONEWORKER)
+          case _ =>
+              resourceRequests = resourceRequests :+ ResourceRequest(Resource(tasks.size))
+        }
+      }
+    })
+    resourceRequests
+  }
+
+  def resourceAllocated(workerId : Int, executorId: Int) : Option[TaskLaunchData] = {
+    val locality = WorkerLocality(workerId)
+
+    // first try to schedule with worker locality. If not found, search tasks without locality
+    val task = scheduleTaskWithLocality(locality).orElse(scheduleTaskWithLocality(NonLocality))
+    task.map(task => addTaskToExecutor(executorId, task.taskId))
+    task
+  }
+
+  private def addTaskToExecutor(executorId: Int, task: TaskId): Unit = {
+    var taskSetForExecutorId = executorToTaskIds.getOrElse(executorId, Set.empty[TaskId])
+    taskSetForExecutorId += task
+    executorToTaskIds += executorId -> taskSetForExecutorId
+  }
+
+  private def scheduleTaskWithLocality(locality: Locality): Option[TaskLaunchData] = {
+    tasksToBeScheduled.get(locality).flatMap { queue =>
+      if (queue.isEmpty) {
+        None
+      } else {
+        Some(queue.dequeue())
+      }
+    }
+  }
+
+  def executorFailed(executorId: Int) : Array[ResourceRequest] = {
+
+    val tasksForExecutor = executorToTaskIds.get(executorId)
+    executorToTaskIds = executorToTaskIds - executorId
+
+    tasksForExecutor match {
```

Comment:
I have comments in https://github.com/intel-hadoop/gearpump/pull/403#discussion_r24474094
that I think it is more readble to do case match for Option when the operation is relative complex(cannot be expressed in single line)